File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <string>
#include <math.h>
#include <vector>
using namespace std;

int solution(string dartResult) {
    int answer = 0;
    vector<int> scores(3, 0);
    int idx = -1;
    
    for(int i = 0 ; i < dartResult.size() ; i++) {
        
        char cur = dartResult[i];
        
        if(isdigit(cur)){
            idx++;
            if(cur == '1' && dartResult[i + 1] == '0'){
                scores[idx] = 10;
                i++;
            }else{
                scores[idx] = cur - '0';
            }
        }else if(cur == 'D')
            scores[idx] = pow(scores[idx], 2);
        else if(cur == 'T')
            scores[idx] = pow(scores[idx], 3);
        else if(cur == '#')
            scores[idx] *= -1;
        else if(cur == '*'){
            scores[idx] *= 2;
            if(idx > 0)
                scores[idx - 1] *= 2; 
        }
        
        
        
        
    }
    
    for(auto s : scores)
        answer += s;
    
    return answer;
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(String dartResult) {
        
        List<Integer> list = new ArrayList<>(); // 숫자 저장
        
        String prev = "";
        for(char c : dartResult.toCharArray()) {
            
            // 싱글 더블 트리플
            if(c == 'S' || c == 'D' || c == 'T'){
                int val;
                if(c == 'S')
                    val = 1;
                else if(c == 'D')
                    val = 2;
                else
                    val = 3;
                list.add((int)Math.pow(Integer.valueOf(prev), val));
                prev = "";
                continue;
            }
            
            // *이거나 #일 경우
            if(c == '*' || c == '#'){
                int len = list.size();
                if(c == '#'){
                    list.set(len - 1, list.get(len - 1) * -1);
                }
                else if(c == '*'){
                    if(len == 1)
                        list.set(0, list.get(0) * 2);
                    else
                        for(int i = 1 ; i <= 2 ; i++)
                            list.set(len - i, list.get(len - i) * 2);
                }
                
                continue;
            }
            
            // 숫자일 경우
            prev += c;
        }
        
        return (int)list.stream().mapToInt(Integer::valueOf).sum();
        
    }
}